src/getINP.ts

```diff
@@ -14,118 +14,127 @@
  * limitations under the License.
  */
 
+import {onBFCacheRestore} from './lib/bfcache.js';
 import {bindReporter} from './lib/bindReporter.js';
 import {initMetric} from './lib/initMetric.js';
-import {observe, PerformanceEntryHandler} from './lib/observe.js';
-import {onBFCacheRestore} from './lib/onBFCacheRestore.js';
+import {observe} from './lib/observe.js';
 import {onHidden} from './lib/onHidden.js';
-import {PerformanceEventTiming, ReportHandler} from './types.js';
+import {getInteractionCount, initInteractionCountPolyfill} from './lib/polyfills/interactionCountPolyfill.js';
+import {Metric, PerformanceEventTiming, ReportHandler} from './types.js';
 
-/*
- * In order to compute a High Percentile (p98-p100) Interaction for INP,
- * we need to store a list of the worst interactions measured.
- * 
- * EVERY_N is the number of interactions before moving to the next-highest (i.e. p98)
- * NUM_ENTRIES_TO_STORE is the max size of the list of entries
- * 
- * EVERY_N * NUM_ENTRIES_TO_STORE becomes, effectively, the max number of interactions
- * per page load for which getINP() works well.  Adjust as needed.
- */
-const EVERY_N = 50;
-const NUM_ENTRIES_TO_STORE = 10;
-const largestINPEntries: PerformanceEventTiming[] = [];
-let minKnownInteractionId = Number.POSITIVE_INFINITY;
-let maxKnownInteractionId = 0;
-
-function updateInteractionIds(interactionId: number): void {
-	minKnownInteractionId = Math.min(minKnownInteractionId, interactionId);
-	maxKnownInteractionId = Math.max(maxKnownInteractionId, interactionId);
+interface Interaction {
+  id: number;
+  latency: number;
+  entries: PerformanceEventTiming[];
 }
 
-function estimateInteractionCount(): number {
-  return (maxKnownInteractionId > 0) ? ((maxKnownInteractionId - minKnownInteractionId) / 7) + 1 : 0;
-}
+// Used to store the interaction count after a bfcache restore, since p98
+// interaction latencies should only consider the current navigation.
+let prevInteractionCount = 0;
+
+// To prevent unnecessary memory usage on pages with lots of interactions,
+// store at most 10 of the longest interactions to consider as INP candidates.
+const MAX_INTERACTIONS_TO_CONSIDER = 10;
+
+// A list of longest interactions on the page (by latency) sorted so the
+// longest one is first. The list is as most MAX_INTERACTIONS_TO_CONSIDER long.
+let longestInteractions: Interaction[] = [];
+
+/**
+ * Takes a performance entry and adds it to the list of worst interactions
+ * if its duration is long enough to make it among the worst. If the
+ * entry is part of an existing interaction, it is merged and the latency
+ * and entries list is updated as needed.
+ */
+const processEntry = (entry: PerformanceEventTiming) => {
+  // Only process the entry if it's possibly one of the ten longest.
+  if (longestInteractions.length < MAX_INTERACTIONS_TO_CONSIDER ||
+      entry.duration > longestInteractions[longestInteractions.length - 1].latency) {
```

Comment:
Ahh, good catch. I think we want to keep the optimization (since it contains a `findIndex()` and `sort()`) but add an additional check to see if the entry is part of an existing interaction. I think that covers all cases, right?

Reply:
I think so!

Reply:
Done.